### PR TITLE
[RISCV] Be more aggressive about forming floating point constants

### PR DESCRIPTION
**Pull request by preames**
We were being very conservative about forming floating point constants via an integer materialization sequence and a fmv.  With the default threshold of 2, we'd only do so if the bit sequence could be produced with a single instruction (LUI, ADDI, or sometimes BSETI).

This change removes the separate threshold entirely, and ties the floating point expansion costing to the integer costing threshold.  The effect of this is that the default threshold increases by 2, and that more sequences are materialized via integers - avoiding constant pool loads.  This is sufficient to cover all constants of types fp16, bf16, and fp32.  Many f64 constants are covered as well, but not all.

One downside of this change is that double constants for ELEN=64 configurations on rv32 can't be formed via the same integer sequences.  This causes a forking in the test coverage which is more than a tad ugly.  Ideas on how to reduce this, or restructure tests to avoid it are more than welcome.

### CODE REVIEW

**Comment by asb** [comment]
A load directly to FPR (especially if you can assume it's likely cached) vs a GPR to FPR move isn't trivial to reason about and of course is very microarch dependent, but this is probably a sensible default. @topperc - any insight on how this would impact the SiFive microarchitectures?

**Comment by preames** [comment]
Throwing this out there for consideration.  

The vast majority of the vector test diff duplication are for a single double constant: 4503599627370496.  This number happens to be precisely representable as a single precision value as well, and thus we could use the sequence.  LUI + FMV.S.X + FCVT.D.S to form it.  If we preferred this form on both rv32 and rv64, it'd make the test diffs collapse.

Not sure that's a good idea though as the cost of the FMV + FCVT may be higher than the slli + fmv we'd use in this change.

**Comment by preames** [comment]
ping

File: llvm/test/CodeGen/RISCV/bfloat-imm.ll [hunk 1]
```diff
@@ -1,14 +1,15 @@
 ; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 2
 ; RUN: llc -mtriple=riscv32 -mattr=+experimental-zfbfmin -verify-machineinstrs \
-; RUN:   -target-abi ilp32f < %s | FileCheck %s
+; RUN:   -target-abi ilp32f < %s | FileCheck --check-prefixes=CHECK %s
```
Comment (topperc) [inline llvm/test/CodeGen/RISCV/bfloat-imm.ll:3]: Why do we need to add --check-prefixes here?

File: llvm/lib/Target/RISCV/RISCVISelLowering.cpp [hunk 2]
```diff
@@ -2031,10 +2025,10 @@ bool RISCVTargetLowering::isFPImmLegal(const APFloat &Imm, EVT VT,
...
   // Building an integer and then converting requires a fmv at the end of
   // the integer sequence.
-  const int Cost =
-    1 + RISCVMatInt::getIntMatCost(Imm.bitcastToAPInt(), Subtarget.getXLen(),
-                                   Subtarget.getFeatureBits());
-  return Cost <= FPImmCost;
+  int64_t Val = Imm.bitcastToAPInt().getSExtValue();
+  RISCVMatInt::InstSeq Seq =
+      RISCVMatInt::generateInstSeq(Val, Subtarget.getFeatureBits());
+  return Seq.size() + 1 <= Subtarget.getMaxBuildIntsCost();
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVISelLowering.cpp:2031]: Should we drop the +1 for Z*inx?

File: llvm/test/CodeGen/RISCV/bfloat-imm.ll (hunk 1 above)
Comment (preames) [reply llvm/test/CodeGen/RISCV/bfloat-imm.ll:3]: We don't.  I think this was a bad rebase.  New version pending shortly.

File: llvm/lib/Target/RISCV/RISCVISelLowering.cpp [hunk 3]
```diff
@@ -8920,6 +8914,17 @@ SDValue RISCVTargetLowering::lowerINSERT_SUBVECTOR(SDValue Op,
       return DAG.getBitcast(Op.getValueType(), SubVec);
     }
 
+    // Shrink down Vec so we're performing the slideup on a smaller LMUL.
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVISelLowering.cpp:8917]: I think you got an extra commit in here?

**Comment by preames** [comment]
Abandoning due to lack of progress in review and low priority.

**Pull request closed without merging**